### ISSUE TITLE
Huggingface trl implementation code

### ISSUE THREAD

**Issue by taemin6697** [facebookresearch/LayerSkip#32]
I looked at the code in Hugging Face TRL, and it seems that a feature like layer dropout is not implemented. Is it still in an unimplemented state?

**Comment by jiogenes** [comment]
Unfortunately, yes... They haven't implemented training code. Refer #1

**Comment by taemin6697** [comment]
from typing import Optional, Union

import torch
from torch import nn
import math

from transformers import LlamaForCausalLM, Cache, LlamaConfig, ROPE_INIT_FUNCTIONS, DynamicCache
from transformers.masking_utils import create_causal_mask
from transformers.modeling_flash_attention_utils import FlashAttentionKwargs
from transformers.modeling_outputs import CausalLMOutputWithPast, BaseModelOutputWithPast
from transformers.models.llama.modeling_llama import KwargsForCausalLM, LlamaModel
from transformers.processing_utils import Unpack
from transformers.utils import logging

logger = logging.get_logger(__name__)


class EE_LlamaModel(LlamaModel):
    def __init__(self, config: LlamaConfig):
        super().__init__(config)
        p_max = getattr(self.config, "p_max", 0.2)
        num_layers = self.config.num_hidden_layers

        # LayerSkip 논문의 지수적(exponential) 확률 공식을 적용합니다.
        self.layer_dropout_probs = torch.zeros(num_layers)
        for l in range(num_layers):
            # D(l) = exp(l * ln(2) / (L-1)) - 1, 이 스케일에 p_max를 곱합니다.
            # 논문 공식은 0~1 스케일을 만들고, 여기에 p_max를 곱해 최종 확률을 정합니다.
            prob = p_max * (math.exp(l * math.log(2) / (num_layers - 1)) - 1)
            self.layer_dropout_probs[l] = prob

    def forward(
            self,
            input_ids: Optional[torch.LongTensor] = None,
            attention_mask: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.LongTensor] = None,
            start_indices: Optional[torch.LongTensor] = None,
            past_key_values: Optional[Cache] = None,
            inputs_embeds: Optional[torch.FloatTensor] = None,
            use_cache: Optional[bool] = None,
            output_attentions: Optional[bool] = None,
            output_hidden_states: Optional[bool] = None,
            cache_position: Optional[torch.LongTensor] = None,
            early_exit_layer: Optional[int] = None,
            training_step: Optional[int] = None,
            **flash_attn_kwargs: Unpack[FlashAttentionKwargs],
    ) -> BaseModelOutputWithPast:

        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
        output_hidden_states = (
            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
        )
        use_cache = use_cache if use_cache is not None else self.config.use_cache

        if (input_ids is None) ^ (inputs_embeds is not None):
            raise ValueError("You must specify exactly one of input_ids or inputs_embeds")

        if self.gradient_checkpointing and self.training and use_cache:
            logger.warning_once(
                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`."
            )
            use_cache = False

        # TODO (joao): remove this exception in v4.56 -- it exists for users that try to pass a legacy cache
        if not isinstance(past_key_values, (type(None), Cache)):
            raise ValueError("The `past_key_values` should be either a `Cache` object or `None`.")

        if inputs_embeds is None:
            inputs_embeds = self.embed_tokens(input_ids)

        if use_cache and past_key_values is None:
            past_key_values = DynamicCache()

        if cache_position is None:
            past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
            cache_position = torch.arange(
                past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device
            )

        if position_ids is None:
            position_ids = cache_position.unsqueeze(0)

        causal_mask = create_causal_mask(
            config=self.config,
            input_embeds=inputs_embeds,
            attention_mask=attention_mask,
            cache_position=cache_position,
            past_key_values=past_key_values,
            position_ids=position_ids,
        )

        hidden_states = inputs_embeds

        # create position embeddings to be shared across the decoder layers
        position_embeddings = self.rotary_emb(hidden_states, position_ids)
        # decoder layers
        all_hidden_states = () if output_hidden_states else None
        all_self_attns = () if output_attentions else None

        if early_exit_layer is not None:
            # 값이 주어지면 해당 레이어까지만 선택
            if early_exit_layer > len(self.layers)-1:
                print('Max early_exit_layer is',len(self.layers)-1,'must range in',0,'~',len(self.layers)-1)
                exit()
            layers_to_run = self.layers[:early_exit_layer]
        else:
            # 값이 없으면 모든 레이어 선택
            layers_to_run = self.layers

        for layer_idx, decoder_layer in enumerate(layers_to_run):

            if output_hidden_states:
                all_hidden_states += (hidden_states,)

            if self.training and torch.rand(1).item() < self.layer_dropout_probs[layer_idx]:
                continue

            layer_outputs = decoder_layer(
                hidden_states,
                attention_mask=causal_mask,
                position_ids=position_ids,
                past_key_value=past_key_values,
                output_attentions=output_attentions,
                use_cache=use_cache,
                cache_position=cache_position,
                position_embeddings=position_embeddings,
                **flash_attn_kwargs,
            )

            hidden_states = layer_outputs[0]

            if output_attentions:
                all_self_attns += (layer_outputs[1],)

        hidden_states = self.norm(hidden_states)

        # add hidden states from the last decoder layer
        if output_hidden_states:
            all_hidden_states += (hidden_states,)

        return BaseModelOutputWithPast(
            last_hidden_state=hidden_states,
            past_key_values=past_key_values if use_cache else None,
            hidden_states=all_hidden_states,
            attentions=all_self_attns,
        )


class EE_LlamaForCausalLM(LlamaForCausalLM):
    def __init__(self, config):
        super().__init__(config)
        self.model = EE_LlamaModel(config=config)

    def forward(
            self,
            input_ids: Optional[torch.LongTensor] = None,
            attention_mask: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.LongTensor] = None,
            past_key_values: Optional[Cache] = None,
            inputs_embeds: Optional[torch.FloatTensor] = None,
            labels: Optional[torch.LongTensor] = None,
            use_cache: Optional[bool] = None,
            output_attentions: Optional[bool] = None,
            output_hidden_states: Optional[bool] = None,
            cache_position: Optional[torch.LongTensor] = None,
            logits_to_keep: Union[int, torch.Tensor] = 0,
            early_exit_layer: Optional[int] = None,
            training_step: Optional[int] = None,
            total_training_steps: Optional[int] = None,
            **kwargs: Unpack[KwargsForCausalLM],
    ) -> CausalLMOutputWithPast:
        r"""
        labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
            Labels for computing the masked language modeling loss. Indices should either be in `[0, ...,
            config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored
            (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
        """
        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
        output_hidden_states = (
            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
        )
        # decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)
        # print(input_ids)
        outputs: BaseModelOutputWithPast = self.model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            position_ids=position_ids,
            past_key_values=past_key_values,
            inputs_embeds=inputs_embeds,
            use_cache=use_cache,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            cache_position=cache_position,
            early_exit_layer=early_exit_layer,
            training_step=training_step,
            **kwargs,
        )

        # 모델의 모든 레이어 출력과 최종 레이어 출력을 변수에 저장
        all_hidden_states = outputs.hidden_states
        final_hidden_state = all_hidden_states[-1]
        logits = self.lm_head(final_hidden_state)

        loss = None
        if labels is not None and self.training:
            num_layers = self.config.num_hidden_layers
            e_scale = getattr(self.config, "e_scale", 1.0)

            # --- 가중치 및 커리큘럼 계산 (여기까지는 동일) ---
            unnormalized_weights = torch.zeros(num_layers, device=labels.device)
            if num_layers > 1:
                for l in range(num_layers - 1):
                    #unnormalized_weights[l] = e_scale * (l * (l + 1) / 2.0)
                    unnormalized_weights[l] = e_scale * ((l + 1) * (l + 2) / 2.0)
                l_final = num_layers - 1
                unnormalized_weights[l_final] = l_final + e_scale * ((l_final - 1) * l_final / 2.0)

            R = getattr(self.config, "R", 4)
            # training_step은 외부에서 동적으로 받아야 함
            #training_step=300
            if R is not None and training_step is not None:
                #print(training_step)
                curriculum_mask = torch.zeros(num_layers, device=labels.device)
                for l in range(num_layers):
                    if (l % R) == (training_step % R):
                        curriculum_mask[l] = 1.0
                active_weights = unnormalized_weights * curriculum_mask
            else:
                active_weights = unnormalized_weights

            sum_of_active_weights = active_weights.sum()
            if sum_of_active_weights > 0:
                normalized_weights = active_weights / sum_of_active_weights
            else:
                normalized_weights = torch.zeros_like(active_weights)
            # --- 가중치 계산 완료 ---

            # --- VRAM 최적화된 손실 계산 시작 ---
            # 1. 최종 손실을 0으로 초기화합니다.
            final_loss = torch.tensor(0.0, device=labels.device)

            # 2. 루프를 돌며 가중치가 0이 아닌 레이어에 대해서만 손실을 계산하고 바로 더합니다.
            for l in range(num_layers):
                weight = normalized_weights[l]
                # 가중치가 0이면 굳이 비싼 연산을 할 필요가 없습니다.
                if weight > 0:
                    hidden_state = all_hidden_states[l + 1]
                    layer_logits = self.lm_head(hidden_state)
                    # CrossEntropyLoss는 logits과 labels의 차원을 맞춰줘야 합니다.
                    layer_loss = self.loss_function(layer_logits.view(-1, self.vocab_size), labels.view(-1),vocab_size=self.config.vocab_size ,**kwargs)
                    #layer_loss = self.loss_function(layer_logits, labels)

                    # 계산된 손실에 가중치를 곱해 바로 누적합니다.
                    final_loss += layer_loss * weight

            loss = final_loss
            # --- 손실 계산 종료 ---

        elif labels is not None:
            # 평가 시에는 최종 로짓으로만 손실을 계산합니다.
            loss = self.loss_function(logits.view(-1, self.vocab_size), labels.view(-1), vocab_size=self.config.vocab_size, **kwargs)

        return CausalLMOutputWithPast(
                    loss=loss,
                    logits=logits,
                    past_key_values=outputs.past_key_values,
                    hidden_states=outputs.hidden_states,
                    attentions=outputs.attentions,
                )

if __name__ == '__main__':
    # --- How to load a pretrained model ---
    # In a real scenario, you would load a pretrained model and its config automatically.
    # NOTE: For this to work, you must first register the custom class with the AutoModel classes.

    # 1. Register your custom model class. This tells `AutoModelForCausalLM` how to load it.
    model_path = "../huggingface_model/Llama-3.2-3B-Instruct"  # Example path

    print(f"--- Loading model from: {model_path} ---")
    print("NOTE: The following `from_pretrained` call is commented out as it requires a real model path.")
    print("\n--- Creating a dummy model for demonstration ---")

    model = EE_LlamaForCausalLM.from_pretrained(model_path)
    config = model.config

    print("Dummy model created successfully.")

    # --- 1. Training Example ---
    model.train()  # Set the model to training mode
    print("\n--- Simulating Training Pass ---")
    print("Layer Dropout and Early Exit Loss are now ACTIVE.")
    dummy_input_ids = torch.randint(0, config.vocab_size, (2, 10))
    dummy_labels = torch.randint(0, config.vocab_size, (2, 10))
    training_outputs = model(input_ids=dummy_input_ids, labels=dummy_labels, output_hidden_states=True)
    training_loss = training_outputs.loss
    print(f"Calculated training loss: {training_loss.item():.4f}")

    # --- 2. Inference Example (Early Exit) ---
    model.eval()  # Set the model to evaluation mode
    print("\n--- Simulating Inference Pass with Early Exit ---")
    print("Layer Dropout and Early Exit Loss are now INACTIVE.")
    with torch.no_grad():
        inference_outputs = model(input_ids=dummy_input_ids, early_exit_layer=4, output_hidden_states=True)
        logits = inference_outputs.logits
    print(f"Inference performed with early exit at layer 4.")
    print(f"Output logits shape: {logits.shape}")

**Comment by taemin6697** [comment]
I've done some of the implementation, and it works well as a result. This code might not be a perfect implementation, but I think it's similar.

[model.txt](https://github.com/user-attachments/files/21653101/model.txt)
[model_grad.txt](https://github.com/user-attachments/files/21653100/model_grad.txt)
